pkg/pillar/cmd/zedagent/zedagent.go

```diff
@@ -293,323 +439,788 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject) in
-		AgentName: agentName,
-		TopicType: types.NetworkXObjectConfig{},
-	})
-	if err != nil {
-		log.Fatal(err)
-	}
-	getconfigCtx.pubNetworkXObjectConfig = pubNetworkXObjectConfig
+func waitUntilZbootReady(zedagentCtx *zedagentContext, stillRunning *time.Ticker) {
+	getconfigCtx := zedagentCtx.getconfigCtx
+	zedagentCtx.subOnboardStatus.Activate()
+	zedagentCtx.subZbootStatus.Activate()
+	getconfigCtx.subNodeAgentStatus.Activate()
+	for !zedagentCtx.zbootRestarted {
+		select {
+		case change := <-zedagentCtx.subOnboardStatus.MsgChan():
+			zedagentCtx.subOnboardStatus.ProcessChange(change)
 
-	pubNetworkInstanceConfig, err := ps.NewPublication(pubsub.PublicationOptions{
-		AgentName: agentName,
-		TopicType: types.NetworkInstanceConfig{},
-	})
-	if err != nil {
-		log.Fatal(err)
-	}
-	getconfigCtx.pubNetworkInstanceConfig = pubNetworkInstanceConfig
+		case change := <-zedagentCtx.subZbootStatus.MsgChan():
+			zedagentCtx.subZbootStatus.ProcessChange(change)
+			if zedagentCtx.zbootRestarted {
+				log.Functionf("Zboot reported restarted")
+			}
 
-	pubAppInstanceConfig, err := ps.NewPublication(pubsub.PublicationOptions{
-		AgentName: agentName,
-		TopicType: types.AppInstanceConfig{},
-	})
-	if err != nil {
-		log.Fatal(err)
-	}
-	getconfigCtx.pubAppInstanceConfig = pubAppInstanceConfig
-	pubAppInstanceConfig.ClearRestarted()
+		case change := <-getconfigCtx.subNodeAgentStatus.MsgChan():
+			getconfigCtx.subNodeAgentStatus.ProcessChange(change)
 
-	pubAppNetworkConfig, err := ps.NewPublication(pubsub.PublicationOptions{
-		AgentName: agentName,
-		TopicType: types.AppNetworkConfig{},
-	})
-	if err != nil {
-		log.Fatal(err)
+		case <-stillRunning.C:
+			// Fault injection
+			if zedagentCtx.fatalFlag {
+				log.Fatal("Requested fault injection to cause watchdog")
+			}
+		}
+		if zedagentCtx.hangFlag {
+			log.Functionf("Requested to not touch to cause watchdog")
+		} else {
+			zedagentCtx.ps.StillRunning(agentName, warningTime, errorTime)
+		}
 	}
-	pubAppNetworkConfig.ClearRestarted()
-	getconfigCtx.pubAppNetworkConfig = pubAppNetworkConfig
+}
 
-	// XXX defer this until we have some config from cloud or saved copy
-	pubAppInstanceConfig.SignalRestarted()
+func waitUntilDNSReady(zedagentCtx *zedagentContext, stillRunning *time.Ticker) {
+	getconfigCtx := zedagentCtx.getconfigCtx
+	dnsCtx := zedagentCtx.dnsCtx
+	log.Functionf("Waiting until we have DeviceNetworkStatus")
+
+	zedagentCtx.subOnboardStatus.Activate()
+	zedagentCtx.subGlobalConfig.Activate()
+	dnsCtx.subDeviceNetworkStatus.Activate()
+	zedagentCtx.subAssignableAdapters.Activate()
+	zedagentCtx.subDevicePortConfigList.Activate()
+	getconfigCtx.subNodeAgentStatus.Activate()
+	zedagentCtx.subVaultStatus.Activate()
+	zedagentCtx.subAttestQuote.Activate()
+	zedagentCtx.subEncryptedKeyFromDevice.Activate()
+	getconfigCtx.subAppNetworkStatus.Activate()
+	zedagentCtx.subWwanMetrics.Activate()
+	zedagentCtx.subLocationInfo.Activate()
```

Comment:
I find it very odd that a function with the purpose (based on its name) being for DNS has code which activates lots of subscriptions. That isn't what modularity means.

If you don't want all activate calls in Run(), then please define some ActivateAll() function which is called when all of the publications and subscriptions have been created.

Reply:
Upon closer inspection it turned out that all subscriptions (incl. persisted ones) can be activated once the zedagent context is initialized and device is onboarded, aka device UUID is known. This allows to simplify the code quite a bit and avoid having these Activate() calls all over the place.